File: Graph/csharp/ShortestPath.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class ShortestPath
{
    public static int Solve(IEnumerable<(string, string)> edges, string nodeA, string nodeB)
    {
        var graph = GraphUtils.BuildUndirectedGraph(edges);
        var queue = new Queue<(string Node, int Distance)>();
        var visited = new HashSet<string> { nodeA };

        queue.Enqueue((nodeA, 0));

        while (queue.Count > 0)
        {
            var (node, distance) = queue.Dequeue();
            if (node == nodeB)
            {
                return distance;
            }

            if (!graph.TryGetValue(node, out var neighbors))
            {
                continue;
            }

            foreach (var neighbor in neighbors)
            {
                if (visited.Add(neighbor))
                {
                    queue.Enqueue((neighbor, distance + 1));
                }
            }
        }

        return -1;
    }
}

File: Graph/python/shortestPath.py
from collections import deque
from typing import Deque, List, Tuple

from .graph_utils import Graph, build_undirected_graph


def shortest_path(edges: List[Tuple[str, str]], node_a: str, node_b: str) -> int:
    graph = build_undirected_graph(edges)
    queue: Deque[tuple[str, int]] = deque([(node_a, 0)])
    visited = {node_a}

    while queue:
        node, distance = queue.popleft()
        if node == node_b:
            return distance
        for neighbor in graph.get(node, []):
            if neighbor in visited:
                continue
            visited.add(neighbor)
            queue.append((neighbor, distance + 1))
    return -1
